[GOVCMS-4390]: Chunk govcms-deploy.

steveworley opened the pull request:
This should provide small testable pieces of the govcms-deploy script which can be
ordered in the .lagoon.yml file. This will allow PaaS users to order their deployment
process but also use supported scripts.

It will allow us to more quickly reorganise the scripts if we need to change the
deployment process for a particular release.

tobybellwood commented on scripts/deploy/db-sync:
```diff
@@ -0,0 +1,35 @@
+BOOTSTRAP=$(drush status --fields=bootstrap | grep -c "Successful")
+
+# Synchronise the environment on the first deploy.
+if [ "$BOOTSTRAP" -eq 1 ] && [ "$GOVCMS_DEPLOY_WORKFLOW_CONTENT" != "import" ]; then
```
Is this overrideable - e.g. for canary/test environments, there is merit in making them sync from prod regularly to "stay fresh"?

steveworley replied:
Yeah definitely agree - if we set `GOVCMS_DEPLOY_WORKFLOW_CONTENT` on the canary environment to `import` it should always pull back a fresh db.

tobybellwood commented on scripts/deploy/pre-updb:
```diff
@@ -0,0 +1,31 @@
+echo "GovCMS Deploy :: Pre-deploy database updates"
+echo "Environment type: $LAGOON_ENVIRONMENT_TYPE"
+
+drush updatedb -y
```
is there scope to broaden this to insert a command in this step from an env_var? - in most cases it may be an updb, but it could also equally be a drush pmu/en or a drush cset?

steveworley replied:
I think I like the flexibility of that, I'm not sure that being able to execute arbitrary commands fits the atomic pattern here. I've added a new `pre-deploy-commands` script which we should probably discuss before including, but basically what this will do is.

- Look for an env var defined as `GOVCMS_DEPLOY_PRE_COMMANDS`
- If that is not defined it'll use `$1`
- If neither are defined it will skip
- It will match the command given on a whitelist of drush commands that we will allow for example it's currently matching on `^(pmu|pm:uninstall|config:set|cset).*`

tobybellwood replied:
nice!

tobybellwood commented on scripts/deploy/govcms-enable-modules:
```diff
@@ -0,0 +1,46 @@
+# Enable Lagoon required modules in production.
+for MODULE in "${PLATFORM_MODULES[@]}"; do
+    if [[ $(echo "$MODULES" | grep -c "$MODULE") -eq 0 ]]; then
+    drush pm:enable "$MODULE" -y
```
is there a time difference between enabling them in a loop, or as a single string list?

steveworley replied:
I was actually thinking that - I'll try and time it and see how we go.

steveworley replied:
```
Successfully enabled: stage_file_proxy, redis, fast404, clamav, robotstxt, lagoon_logs, environment_indicator
real	0m7.263s
user	0m4.259s
sys	0m1.022s
```

```
 [success] Successfully enabled: stage_file_proxy
 [success] Successfully enabled: redis
 [success] Successfully enabled: fast404
 [success] Successfully enabled: clamav
 [success] Successfully enabled: robotstxt
 [success] Successfully enabled: lagoon_logs
 [success] Successfully enabled: environment_indicator

real	0m19.370s
user	0m11.471s
sys	0m3.117s
```

steveworley replied:
Enabling in one go is a winner - updated.

tobybellwood commented on scripts/deploy/govcms-enable-modules:
```diff
@@ -0,0 +1,46 @@
+)
+
+if [ "$LAGOON_ENVIRONMENT_TYPE" != "production" ]; then
+  for MODULE in "${NON_PROD_MODULES[@]}"; do
```
do we really not want the PLATFORM_MODULES enabled in non-prod?

steveworley replied:
This was just for parity with govcms_lagoon (it's enabled on both I believe - will double check).

simesy replied:
FWIW, my preferred is to leave it installed (avoid install/uninstalling Drupal modules as a general rule) and do this for production sites in *.settings.php
```
      $config['stage_file_proxy.settings']['origin'] = NULL;
```

steveworley replied:
While I agree in principle - there are some modules that we need to ensure are _always_ enabled for security and policy reasons. If configuration management is up to a site to manage then they will be able to disable these modules.

I think this might warrant a bigger discussion about how we want to mandate platform requirements.

tobybellwood commented on composer.json:
```diff
@@ -25,6 +25,16 @@
+        "scripts/deploy/govcms-database-update",
+        "scripts/deploy/govcms-enable-modules",
+        "scripts/deploy/govcms-pre-deploy-commands",
+        "scripts/deploy/govcms-pre-deploy-update-datase",
```
```suggestion
        "scripts/deploy/govcms-pre-deploy-update-database",
```

tobybellwood commented on composer.json:
```diff
@@ -25,6 +25,16 @@
+        "scripts/deploy/govcms-cache-rebuild",
+        "scripts/deploy/govcms-config-import",
+        "scripts/deploy/govcms-db-sync",
+        "scripts/deploy/govcms-database-update",
```
```suggestion
        "scripts/deploy/govcms-update-database",
```

tobybellwood commented on composer.json:
```diff
@@ -25,6 +25,16 @@
         "scripts/govcms-lint-distro",
         "scripts/govcms-phpunit",
         "scripts/govcms-vet",
-        "scripts/govcms-deploy"
```
if we leave govcms-deploy in there for a transition period, we can get this merged faster, no?

steveworley replied:
Quite, should deprecate over backwards breaking changes :D

simesy commented on scripts/deploy/govcms-prepare:
```diff
@@ -0,0 +1,12 @@
+# sites are in a deployable position.
+#
+
+sed -i "s/%%PROJECT_NAME%%/\${LAGOON_PROJECT}/g" /app/drush/sites/govcms.site.yml
```
If all sites are using the latest image then this script might now be redundant. The alias works differently now.
https://github.com/govCMS/govcms8lagoon/blob/develop/.docker/images/govcms8/govcms.site.yml

So making a whole script for this one deprecated line is too much IMO

steveworley replied:
Yeah - agree I didn't really audit them as I moved them over. Probably should have I was going for parity first and then hoping we could revist and update!

simesy replied:
Yeah agreed, just a deprecation notice on the while script means we capture the knowledge at least.

simesy commented:
I love this.

AlexSkrypnyk commented:
One thing that requires attention is the naming of the commands. It is currently inconsistent.

The format should be either 1) `govcms-[action]-[object]` or 2)`govcms-[object]-[action]`. Drush uses 2), so it is probably better to stick with it (I personally prefer 1) ).

So, for 2):

```
        "scripts/deploy/govcms-backup-config", - not OK - should be `govcms-config-backup`
        "scripts/deploy/govcms-backup-db", - not OK - should be `govcms-db-backup`
        "scripts/deploy/govcms-cache-rebuild", - OK
        "scripts/deploy/govcms-config-import", - OK
        "scripts/deploy/govcms-db-sync", - OK
        "scripts/deploy/govcms-update-database", -  not OK - should be `govcms-db-update`
        "scripts/deploy/govcms-enable-modules", -  not OK - should be `govcms-enable_modules`
        "scripts/deploy/govcms-pre-deploy-commands", - why do we have "commands" in the name and not just "govcms-pre-deploy". In either cases - it is not clear what this script does.
        "scripts/deploy/govcms-pre-deploy-update-database" - not OK - `govcms-pre-deploy-db-update`
        "scripts/deploy/govcms-prepare", - not clear what this means
        "scripts/deploy/govcms-preserve-backups" - not OK - `govcms-backups-preserve`
```

AlexSkrypnyk commented:
@steveworley the CI fails because files are not passing shellcheck. Can you please make required amendments. Thanks

steveworley commented:
> One thing that requires attention is the naming of the commands. It is currently inconsistent.

Yeah agreed I've updated to reflect that with the latest commit. Some key things to take away:

- I've renamed `govcms-prepare` to `govcms-update_site_alias` - this is really a legacy script to not change the govcms-deploy process too much, this should be deprecated and removed
- `pre-deploy-commands` this is to accommodate running commands that we need to via environment variables... I'm not sure I agree with this entirely, but it was a suggestion above.

AlexSkrypnyk commented on scripts/deploy/govcms-backups-preserve:
```diff
@@ -0,0 +1,14 @@
+echo "GovCMS Deploy :: Preserve backups"
+
+if [ -f "$GOVCMS_BACKUP_DIR/pre-deploy-dump.sql.gz" ]; then
+  mv "$GOVCMS_BACKUP_DIR/pre-deploy-dump.sql.gz" "$GOVCMS_BACKUP_DIR/pre-deploy-dump-last-good.sql.gz"
```
@steveworley Do we want to output something like
`Moved $GOVCMS_BACKUP_DIR/pre-deploy-dump.sql.gz to $GOVCMS_BACKUP_DIR/pre-deploy-dump-last-good.sql.gz`

also, do we want to add an output (`else`) if the dump does not exist. 

This way script will be more verbose which is what site owners probably want (so that scripts are not perceived as black boxes).

@tobybellwood @simesy 
What is your opinion on verbosity? (this script is just an example; the question addresses all deploy scripts)

AlexSkrypnyk commented on scripts/deploy/govcms-cache-rebuild:
```diff
@@ -0,0 +1,23 @@
+
+echo "GovCMS Deploy :: Cache rebuild"
+
+if [ "$GOVCMS_CACHE_REBUILD" != "true" ]; then
```
do we want the comparison to the string `"true"` or just to check if the value of the variable is set? Do we have an agreement on this standard across scripts?

AlexSkrypnyk commented on scripts/deploy/govcms-config-backup:
```diff
@@ -0,0 +1,32 @@
+set -euo pipefail
+
+#
+# GovCMS database configuration.
```
incorrect heading - this is a config backup

AlexSkrypnyk commented on scripts/deploy/govcms-config-backup:
```diff
@@ -0,0 +1,32 @@
+
+drush config:export sync -y --destination "$GOVCMS_BACKUP_DIR/config"
+tar -czf "$GOVCMS_BACKUP_DIR/$GOVCMS_CONFIG_BACKUP.tar.gz" -C "$GOVCMS_BACKUP_DIR/config" --remove-files
+
```
probably need to add an output with location of the saved backup

AlexSkrypnyk commented on scripts/deploy/govcms-config-import:
```diff
@@ -0,0 +1,45 @@
+fi
+
+# Check that there are configuration files.
+set +e #
```
do not need the `#` at the end

AlexSkrypnyk commented on scripts/deploy/govcms-config-import:
```diff
@@ -0,0 +1,45 @@
+dev_config_count=$(ls -1 /app/config/dev/*.yml 2>/dev/null | wc -l)
+set -e
+
+if [[ "$config_count" -eq 0 && "$dev_config_count" -eq 0 ]]; then
```
can we please stick to one way of writing conditions

```
if [ "$config_count" -eq 0 ] && [ "$dev_config_count" -eq 0 ]; then
```

AlexSkrypnyk commented on scripts/deploy/govcms-config-import:
```diff
@@ -0,0 +1,45 @@
+  echo "[skip]: There is no configuration."
+  exit 0
+fi
+
```
extra empty line

AlexSkrypnyk commented on scripts/deploy/govcms-config-import:
```diff
@@ -0,0 +1,45 @@
+fi
+
+if [ "$LAGOON_ENVIRONMENT_TYPE" != "production" ] && [ "$dev_config_count" -gt 0 ]; then
+  echo "[update]: Import dev configuration partial"
```
```
echo "[update]: Import dev configuration partially."
```

AlexSkrypnyk commented on scripts/deploy/govcms-db-backup:
```diff
@@ -0,0 +1,32 @@
+
+mkdir -p "$GOVCMS_BACKUP_DIR"
+drush sql:dump --gzip --result-file="$GOVCMS_BACKUP_DIR/pre-deploy-dump.sql"
+
```
maybe add an output to know the location of the backup

AlexSkrypnyk commented on scripts/deploy/govcms-db-backup:
```diff
@@ -0,0 +1,32 @@
+
+# Backup processes only happen on production.
+if [ "$LAGOON_ENVIRONMENT_TYPE" != "production" ]; then
+  echo "[skip]: Non-production environment"
```
missing dot at the end

AlexSkrypnyk commented on scripts/deploy/govcms-config-import:
```diff
@@ -0,0 +1,45 @@
+
+
+if [ "$config_count" -gt 0 ]; then
+  echo "[update]: Import site configuration"
```
missing dot at the end

AlexSkrypnyk commented on scripts/deploy/govcms-db-sync:
```diff
@@ -0,0 +1,48 @@
+  exit 0
+fi
+
+echo "[info]: Check that the site can be bootstrapped"
```
missing dot at the end

AlexSkrypnyk commented on scripts/deploy/govcms-db-sync:
```diff
@@ -0,0 +1,48 @@
+  # This will ensure that the databse is synchronised on the first deploy
+  # of an enviornment in lagoon.
+  echo "[info]: Site could not be bootstrapped... syncing."
+  drush --alias-path=/etc/drush/sites sql-sync @"$GOVCMS_SITE_ALIAS" @self -y
```
could you please confirm that this alias path will play nicely with all platform types (SaaS and PaaS). Just in case if the location of the alias is different.

AlexSkrypnyk commented on scripts/deploy/govcms-pre-deploy-db-update:
```diff
@@ -0,0 +1,35 @@
+# (for example with drush status). This provides the deploy agent
+# the ability to perform updates early in the automated flow.
+#
+# The ansible deploy process will add GOVCMS_DEPLOY_PRE_UPDB prior to
```
can we please remove this comment as it is confusing for consumer sites - they do not know what "ansible deploy" is.

AlexSkrypnyk commented on scripts/deploy/govcms-pre-deploy-db-update:
```diff
@@ -0,0 +1,35 @@
+
+if [ "$GOVCMS_DEPLOY_PRE_UPDB" = false ]; then
+  # A pre-deploy update is not required so return early.
+  echo "[skip]: Environment variable not set"
```
can we please change this message to have more context. Maybe something like "Skipping pre-deployment db update" or something along these lines.

AlexSkrypnyk commented on scripts/deploy/govcms-pre-deploy-db-update:
```diff
@@ -0,0 +1,35 @@
+drush updatedb -y
+
+echo "[success]: Completed successfully."
+
```
pls remove new lines

AlexSkrypnyk commented on scripts/deploy/govcms-update_site_alias:
```diff
@@ -0,0 +1,17 @@
+# scaffold
+#
+
+sed -i "s/%%PROJECT_NAME%%/\${LAGOON_PROJECT}/g" /app/drush/sites/govcms.site.yml
```
this will fail for projects that have their alias file located in another place. Maybe worth wrapping this in `if`.

AlexSkrypnyk commented:
@steveworley Can you please not merge this until I have a proper review and testing done. Thanks.

AlexSkrypnyk commented:
This now is super neat and clean. Ready to be merged.
